src/libs/jpegdec/lv_jpegdec.c

```diff
@@ -0,0 +1,342 @@
+/**
+ * @file lv_jpegdec.c
+ *
+ */
+
+/*********************
+ *      INCLUDES
+ *********************/
+#include "../../../lvgl.h"
+#if LV_USE_JPEGDEC
+
+#include "lv_jpegdec.h"
+#include "jpegdec.h"
+#include <stdlib.h>
+
+/*********************
+ *      DEFINES
+ *********************/
+
+#define DECODER_NAME    "JPEGDEC"
+
+/**********************
+ *      TYPEDEFS
+ **********************/
+
+/**********************
+ *  STATIC PROTOTYPES
+ **********************/
+static lv_result_t decoder_info(lv_image_decoder_t * decoder, const void * src, lv_image_header_t * header);
+static lv_result_t decoder_open(lv_image_decoder_t * decoder, lv_image_decoder_dsc_t * dsc);
+static void decoder_close(lv_image_decoder_t * dec, lv_image_decoder_dsc_t * dsc);
+static lv_draw_buf_t * decode_jpeg_data(JPEGIMAGE * jpg, const void * jpeg_data, size_t jpeg_data_size, int iBPP);
+/**********************
+ *  STATIC VARIABLES
+ **********************/
+/**********************
+ *      MACROS
+ **********************/
+
+/**********************
+ *   GLOBAL FUNCTIONS
+ **********************/
+
+/**
+ * Register the JPEG decoder functions in LVGL
+ */
+void lv_jpegdec_init(void)
+{
+    lv_image_decoder_t * dec = lv_image_decoder_create();
+    lv_image_decoder_set_info_cb(dec, decoder_info);
+    lv_image_decoder_set_open_cb(dec, decoder_open);
+    lv_image_decoder_set_close_cb(dec, decoder_close);
+}
+
+void lv_jpegdec_deinit(void)
+{
+    lv_image_decoder_t * dec = NULL;
+    while((dec = lv_image_decoder_get_next(dec)) != NULL) {
+        if(dec->info_cb == decoder_info) {
+            lv_image_decoder_delete(dec);
+            break;
+        }
+    }
+}
+
+/**********************
+ *   STATIC FUNCTIONS
+ **********************/
+
+/**
+ * Get info about a JPEG image
+ * @param decoder   pointer to the decoder where this function belongs
+ * @param src       can be file name or pointer to a C array
+ * @param header    image information is set in header parameter
+ * @return          LV_RESULT_OK: no error; LV_RESULT_INVALID: can't get the info
+ */
+static lv_result_t decoder_info(lv_image_decoder_t * decoder, const void * src, lv_image_header_t * header)
+{
+    LV_UNUSED(decoder); /*Unused*/
+    uint8_t cBuf[32];
+    uint16_t w = 0, h = 0, iMarker = 0;
+    int iBpp;
+    int i = 2, j = 2; /* Offset of first marker */
+    lv_fs_file_t f;
+    lv_fs_res_t res;
+    uint32_t rn, iFileSize;
+    const lv_image_dsc_t * img_dsc = src;
+    lv_image_src_t src_type = lv_image_src_get_type(src);          /*Get the source type*/
+
+    /*If it's a JPEG file...*/
+    if(src_type == LV_IMAGE_SRC_FILE) {
+        const char * fn = src;
+        /* Read the width and height from the file */
+        res = lv_fs_open(&f, fn, LV_FS_MODE_RD);
+        lv_fs_seek(&f, 0, LV_FS_SEEK_END);
+        lv_fs_tell(&f, &rn);
+        lv_fs_seek(&f, 0, LV_FS_SEEK_SET);
+        iFileSize = (int)rn;
+        if(res != LV_FS_RES_OK) return LV_RESULT_INVALID;
```

Comment:
Please move this after `lv_fs_open` and add a log entry. E.g. `LV_LOG_WARN("Can't open %s, result: %d", fn, res);`

Reply:
Updated with both suggested changes